ethcore/src/engines/clique/mod.rs

```diff
@@ -0,0 +1,418 @@
+	}
+
+	fn sign_header(&self, header: &Header) -> Result<Signature, Error> {
+		let digest = sig_hash(header)?;
+		if let Ok(sig) = self.signer.read().sign(digest) {
+			Ok(sig)
+		} else {
+			Err(Box::new("failed to sign header").into())
+		}
+	}
+
+	//pub fn snapshot(self, bn: u64) -> AuthorizationSnapshot {
+		// if we are on a checkpoint block, build a snapshot
+	//}
+}
+
+impl Engine<EthereumMachine> for Clique {
+	fn name(&self) -> &str { "Clique" }
+
+	// nonce + mixHash + extraData
+	fn seal_fields(&self, _header: &Header) -> usize { 2 }
+	fn machine(&self) -> &EthereumMachine { &self.machine }
+	fn maximum_uncle_count(&self, _block: BlockNumber) -> usize { 0 }
+	fn populate_from_parent(&self, header: &mut Header, parent: &Header) {
+		// if in turn, set difficulty
+		//
+		if self.is_signer_proposer(header.number()) {
+				let mut address = Address::new();
+				address.0 = NULL_AUTHOR.clone();
+				header.set_author(address);
+		}
+	}
+
+	fn close_block_extra_data(&self, _header: &Header) -> Option<Vec<u8>> {
+			let mut h = _header.clone();
+
+			if self.is_signer_proposer(_header.number()) {
+				 if let Some(ref mut snapshot) = *self.snapshot.write() {
+					 trace!(target: "engine", "applying sealed block");
+					 snapshot.apply(_header);
+
+					 let signers = &snapshot.signers;
+					 trace!(target: "engine", "applied.  found {} signers", signers.len());
+
+					 let mut v: Vec<u8> = vec![0; SIGNER_VANITY_LENGTH as usize+20*signers.len()+SIGNER_SIG_LENGTH as usize];
+					 let sig_offset = SIGNER_VANITY_LENGTH as usize+20*signers.len();
+					 for i in 0..signers.len() {
+						 //signers[i].copy_to(&v[SIGNER_VANITY_LENGTH as usize+i*20..(i+1)*20]);
+
+						 v[SIGNER_VANITY_LENGTH as usize+i*20..SIGNER_VANITY_LENGTH as usize+(i+1)*20].clone_from_slice(&signers[i]);
+					 }
+
+					 trace!(target: "engine", "writing signature");
+					 h.set_extra_data(v.clone());
+					 v[sig_offset..].copy_from_slice(&self.sign_header(&h).expect("should be able to sign header")[..]);
+					 trace!(target: "engine", "signature written");
+
+					 return Some(v);
+				 } else {
+					 panic!("failed to populate extra data when sealing");
+				 }
+			}
+			return None;
+	}
+
+	fn set_signer(&self, ap: Arc<AccountProvider>, address: Address, password: Password) {
+		self.signer.write().set(ap, address, password);
+	}
+
+	/// None means that it requires external input (e.g. PoW) to seal a block.
+	/// /// Some(true) means the engine is currently prime for seal generation (i.e. node
+	///     is the current validator).
+	/// /// Some(false) means that the node might seal internally but is not qualified
+	///     now.
+	///
+	fn seals_internally(&self) -> Option<bool> {
+		Some(true)
+	}
+
+	/// Attempt to seal generate a proposal seal.
+	///
+	/// This operation is synchronous and may (quite reasonably) not be available, in which case
+	/// `Seal::None` will be returned.
+	fn generate_seal(&self, block: &ExecutedBlock, _parent: &Header) -> Seal {
```

Comment:
The `extraData` logic probably should belong here. So IMO we shouldn't set it in `on_close_block` (that probably wouldn't even work because the engine needs to distinguish authoring and non-authoring clients!).

An easy way to do this is to create another type for enum `engines::Seal` (probably `Seal::ExtraData`), and create a function similar to `LockedBlock::seal`. In this way we can generate seal here.

Reply:
@sorpaas As an alternative, I added a hook `seal_header` which is called at the end of sealing before the block hash is computed: https://github.com/paritytech/parity-ethereum/pull/9981/files#diff-e2e171033e005c3b329818aaff213767R447 .

We are using this as a hook to modify block difficulty (used to weight 'in-turn' signers more heavily so that their block is more likely to be included), and place the header signature in the block extra data.

What do you think about this approach?

Reply:
@jwasinger I still think it might not work. `Block::close` is called by all including those who are not proposers but merely importing blocks. If we call seal header for them the client would panic. Do you mind to explain more why `generate_seal` does not fit this purpose?